repository/repository.go

```diff
@@ -470,3 +476,169 @@ To find more information on specific steps, please see documentation [here](http
+
+	fetchOpts := &git.FetchOptions{
+		RemoteName: remote,
+		RefSpecs:   []config.RefSpec{config.RefSpec(refSpec)},
+		Force:      true,
+	}
+
+	// Fetch the remote branch
+	if err = r.Fetch(fetchOpts); err != nil && err != git.NoErrAlreadyUpToDate {
+		return err
+	}
+
+	// Get the local and remote branch references
+	localRef, err = r.Head()
+	if err != nil {
+		return err
+	}
+
+	remoteRef, err = r.Reference(plumbing.NewRemoteReferenceName(remote, rb), true)
+	if err != nil {
+		return err
+	}
+
+	// Find the commits that are in the local branch but not in the remote branch
+	uniqueCommits, err := findUniqueCommits(r, localRef, remoteRef)
+	if err != nil {
+		return err
+	}
+
+	// Print the commit message and hash for each unique commit
+	for _, commit := range uniqueCommits {
+		fmt.Printf("Commit: %s\nMessage: %s\n\n", commit.Hash, commit.Message)
+	}
+
+	return nil
+}
+
+// PushRemoteBranch will push the local branch to the remote repository
+func PushRemoteBranch(r *git.Repository, remote, u, t string, debug bool) error {
+	var (
+		branchRef string
+		h         *plumbing.Reference
+	)
+
+	h, err := r.Head()
+	if err != nil {
+		return err
+	}
+
+	branchRef = h.Name().String()
+
+	auth := &http.BasicAuth{
+		Username: u,
+		Password: t,
+	}
+
+	opts := &git.PushOptions{
+		RemoteName: remote,
+		Force:      true,
+		RefSpecs:   []config.RefSpec{config.RefSpec(branchRef + ":" + branchRef)},
+		Auth:       auth,
+	}
+
+	if debug {
+		opts.Progress = os.Stdout
+	}
+
+	return r.Push(opts)
+}
+
+// findUniqueCommits returns the commits that are in the first reference but not in the second.
+func findUniqueCommits(r *git.Repository, localRef, remoteRef *plumbing.Reference) ([]*object.Commit, error) {
+	var (
+		localCommits  []*object.Commit
+		remoteCommits []*object.Commit
+		uniqueCommits []*object.Commit
+		err           error
+	)
+
+	// Get the local and remote commits
+	localCommits, err = getCommits(r, localRef.Hash())
+	if err != nil {
+		return nil, err
+	}
+
+	remoteCommits, err = getCommits(r, remoteRef.Hash())
+	if err != nil {
+		return nil, err
+	}
+
+	var remoteCommitMap = make(map[plumbing.Hash]struct{}, len(remoteCommits))
+
+	for _, c := range remoteCommits {
+		remoteCommitMap[c.Hash] = struct{}{}
+	}
+
+	for _, c := range localCommits {
+		if _, found := remoteCommitMap[c.Hash]; !found {
+			uniqueCommits = append(uniqueCommits, c)
+		}
+	}
+
+	return uniqueCommits, nil
+}
+
+// getCommits returns the commits beginning at the given hash
+func getCommits(r *git.Repository, h plumbing.Hash) ([]*object.Commit, error) {
+	var (
+		firstCommit *object.Commit
+		commits     []*object.Commit
+		iter        object.CommitIter
+		err         error
+	)
+
+	firstCommit, err = r.CommitObject(h)
+	if err != nil {
+		return nil, err
+	}
+
+	iter = object.NewCommitPreorderIter(firstCommit, nil, nil)
+
+	limit := 100
+	count := 0
+	err = iter.ForEach(func(c *object.Commit) error {
```

Comment:
This error should be checked for nil first and then check for a certain kind of error.